samples: bluetooth: added throughput many_to_one sample (#26038)
nrfconnect/sdk-nrf: samples/bluetooth/throughput_many_to_one/README.rst

It consists of two roles:

- ``periodic_adv_rsp``: PAwR advertiser/host that creates a periodic advertising set, encodes control frames (open-slot bitmap, ACKs, and retransmit hints), receives responses from multiple devices, and prints measured and theoretical throughput.
- ``periodic_sync_rsp``: PAwR synchronizer/responder that discovers and synchronizes to the periodic advertiser, claims a free response slot, and transmits payloads in its assigned slot.

The synchronizer discovers the advertiser by device name (``"PAwR adv sample"``). The advertiser prints measured and theoretical throughput for the active set of responders.

Default parameter values
========================

.. list-table:: Default parameter values
   :header-rows: 1

   * - Parameter
     - Value
   * - Number of responders (slots)
     - 2
   * - Response payload size
     - 247 bytes
   * - PHY data rate
     - LE 2M
   * - Response slot delay
     - 3 units (3.75 ms)
   * - Response slot spacing
     - computed from payload and PHY
   * - Subevent interval
     - 6 units (7.50 ms)
   * - Retransmit bitmap bytes
     - up to 4

Limitations and trade-offs

> ===========================

Reviewer: This section could be moved towards the bottom, just before the dependencies section.

- **Limited per-response payload**: Each response is capped at 247 bytes by the controller, so larger data must be application-segmented.
- **Control bandwidth limits**: ACK list capacity per event is limited by control payload tailroom, and the retransmit bitmap is truncated to a few bytes to bound control size, delaying hints under heavy load.

These constraints are typical of random access, connectionless schemes. If your use case demands stronger fairness, reliability, or access control, a GATT-based design (optionally using PAST to distribute sync) is usually preferable.

Configuration
**************

|config|

.. _throughput_many_to_one_config_options:

Configuration options
======================

Check and configure the following sample-specific Kconfig options:

.. code-block::

   .. options-from-kconfig::
      :kconfig:option:`CONFIG_BT_NUM_RSP_DEVICES`

.. note::
   Before building the advertiser, set :kconfig:option:`CONFIG_BT_NUM_RSP_DEVICES` in ``periodic_adv_rsp/prj.conf`` to the number of synchronizer devices you plan to run.
   If this value does not match your setup, the advertiser may appear stalled while waiting for unassigned or missing responders.

> 

Reviewer: ```suggestion
   If this value does not match your setup, the advertiser might appear stalled while waiting for unassigned or missing responders.
```

Additional configuration
========================

Key Bluetooth and controller options used by this sample include:

- :kconfig:option:`CONFIG_BT_PER_ADV`
- :kconfig:option:`CONFIG_BT_PER_ADV_RSP`
- :kconfig:option:`CONFIG_BT_PER_ADV_SYNC`
- :kconfig:option:`CONFIG_BT_PER_ADV_SYNC_RSP`
- :kconfig:option:`CONFIG_BT_CTLR_PHY_2M`
- :kconfig:option:`CONFIG_BT_CTLR_DATA_LENGTH_MAX`
- :kconfig:option:`CONFIG_BT_L2CAP_TX_MTU`
- :kconfig:option:`CONFIG_BT_USER_PHY_UPDATE`
- :kconfig:option:`CONFIG_BT_DEVICE_NAME`

Configuration files
===================

This sample provides predefined configuration files for typical use:

- ``periodic_adv_rsp/prj.conf`` — advertiser/host settings (slot count, buffer sizes, timing).

> - ``periodic_sync_rsp/prj.conf`` — synchronizer/responder settings (PAwR sync, payload size, buffers).

Reviewer: ```suggestion
* :file:`periodic_adv_rsp/prj.conf` - Advertiser/host settings (slot count, buffer sizes, and timing).
* :file:`periodic_sync_rsp/prj.conf` - Synchronizer/responder settings (PAwR sync, payload size, and buffers).

```

Building and running
********************

.. |sample path| replace:: :file:`samples/bluetooth/throughput_many_to_one`

.. include:: /includes/build_and_run.txt

Build the two roles separately and program them to different development kits:

Advertiser ::
   cd periodic_adv_rsp
   west build -p always -b nrf54l15dk/nrf54l15/cpuapp 

Synchronizer ::
   cd periodic_sync_rsp
   west build -p always -b nrf54l15dk/nrf54l15/cpuapp 

After building, flash each role to its target kit.

> 

Reviewer: ```suggestion
```

For the synchronizers::

    *** Booting nRF Connect SDK v3.1.99-28569c1ad6cb ***
    *** Using Zephyr OS v4.2.99-c539d475b659 ***
    Starting PAwR Synchronizer
    [SCAN] Waiting for periodic advertiser...
    [SCAN] Creating periodic sync to SID 0
    [SYNC] Synced to 28:0E:0E:0E:D4:A9 (random) with 1 subevents
    [SYNC] Changed sync to subevent 0
    [JOIN] Sent claim for slot 0 with token 0xb4a6d332
    [JOIN] Acked: token 0xb4a6d332 assigned slot 0

.. note::
Retransmission log messages may occur. The advertiser reports missed slots and suggests retransmission, 
but the responder in this sample always sends fresh payloads and does not implement retransmission logic.

Troubleshooting
===============

- [LOSS] slot X missed response; should retransmit
  - The advertiser expected data from that slot this event but didn’t receive it.
  - After first real payload from a slot, retransmit hints may appear under RF loss.

- [CLAIM] Reclaimed idle slot X
  - The advertiser freed a slot after several missed events. Responders should re-join.

- Failed to set subevent data (err -13)
  - Controller rejected PAwR subevent parameters (e.g., subevent index out of range).

>   - Ensure subevent index is modulo the configured number of subevents and data length fits.

Reviewer: ```suggestion
This section lists some common errors and possible solutions to them.

* [LOSS] slot X missed response, retransmit:

  * The advertiser expected data from that slot for this event but did not receive it.
  * After the first real payload from a slot, retransmit hints may appear under RF loss.

* [CLAIM] Reclaimed idle slot X:

  * The advertiser freed a slot after several missed events. 
     Responders must rejoin.

* Failed to set subevent data (err -13):

  * The Controller rejected PAwR subevent parameters (for example, subevent index out of range).
  * Ensure the subevent index is modulo the configured number of subevents and the data length fits.

```

Dependencies
*************

This sample uses the following Zephyr libraries:

* :ref:`zephyr:bluetooth_api`:

  * :file:`include/bluetooth/bluetooth.h`
  * :file:`include/bluetooth/hci.h`

* :ref:`zephyr:kernel_api`:

  * :file:`include/kernel.h`

* :file:`include/zephyr/sys/byteorder.h`
* :file:`include/zephyr/random/random.h`
* :file:`include/zephyr/net/buf.h`
* :file:`include/zephyr/sys/printk.h`
* :file:`include/zephyr/sys/util.h`

It uses the following `sdk-nrfxlib`_ library:

* :ref:`nrfxlib:softdevice_controller`

References
*************

> For more about Periodic Advertising with Responses (PAwR), see the Bluetooth Core Specification:

Reviewer 2: Could this reference be linked?

===============

This section lists some common errors and possible solutions to them.

* [LOSS] slot X missed response; should retransmit:

  * The advertiser expected data from that slot this event but did not receive it.
  * After the first real payload from a slot, retransmit hints may appear under RF loss.

* [CLAIM] Reclaimed idle slot X:

  * The advertiser freed a slot after several missed events.
    Responders must rejoin.

* Failed to set subevent data (err –13):

  * The controller rejected PAwR subevent parameters (for example, subevent index out of range).
  * Ensure subevent index is modulo the configured number of subevents and data length fits.

Limitations and trade-offs
==========================

This sample demonstrates a connectionless, token-based claim/ACK protocol over PAwR.
Consider the following trade-offs compared to GATT connections or using Periodic Advertising Sync Transfer (PAST) with a connected control plane:

* Same-slot collisions - If two unassigned responders choose the same open slot in the same subevent, their claim responses collide at the PHY and the advertiser receives neither.
  No acknowledgment (ACK) is issued that event.

>   Both devicesremain unassigned and retry in later events with probabilistic backoff.

Reviewer: ```suggestion
  Both devices remain unassigned and retry in later events with probabilistic backoff.
```
